overlord/snapstate/refreshhints.go

```diff
@@ -121,3 +135,111 @@ func (r *refreshHints) Ensure() error {
 	}
 	return r.refresh()
 }
+
+func refreshHintsFromCandidates(st *state.State, updates []*snap.Info, ignoreValidationByInstanceName map[string]bool, deviceCtx DeviceContext) ([]*RefreshHint, error) {
+	if ValidateRefreshes != nil && len(updates) != 0 {
+		userID := 0
+		var err error
+		updates, err = ValidateRefreshes(st, updates, ignoreValidationByInstanceName, userID, deviceCtx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	hints := make([]*RefreshHint, len(updates))
+	for _, update := range updates {
+		var snapst SnapState
+		if err := Get(st, update.InstanceName(), &snapst); err != nil {
+			return nil, err
+		}
+		// XXX: is this ok?
+		flags := snapst.Flags
+		flags.IsAutoRefresh = true
+
+		flags, err := ensureInstallPreconditions(st, update, flags, &snapst)
+		if err != nil {
+			// XXX: do we need to handle refreshAll here?
+			logger.Noticef("cannot update %q: %v", update.InstanceName(), err)
+			continue
+		}
+
+		if err := earlyEpochCheck(update, &snapst); err != nil {
+			// XXX: do we need to handle refreshAll here?
+			logger.Noticef("cannot update %q: %v", update.InstanceName(), err)
+			continue
+		}
+		snapsup := &RefreshHint{
+			SnapSetup{
+				Base:         update.Base,
+				Prereq:       defaultContentPlugProviders(st, update),
+				DownloadInfo: &update.DownloadInfo,
+				SideInfo:     &update.SideInfo,
+				Type:         update.Type(),
+				PlugsOnly:    len(update.Slots) == 0,
+				InstanceKey:  update.InstanceKey,
+				auxStoreInfo: auxStoreInfo{
+					Website: update.Website,
+					Media:   update.Media,
+				},
+		}}
+
+		hints = append(hints, snapsup)
+	}
+	return hints, nil
+}
+
+func updatesFromRefreshHints(st *state.State) ([]*RefreshHint, map[string]*SnapState, error) {
+	var hints []*RefreshHint
+	err := st.Get("refresh-hints", &hints)
+	if err != nil && err != state.ErrNoState {
+		return nil, nil, err
+	}
+	if err != nil {
+		return nil, nil, nil
+	}
+
+	snapStates, err := All(st)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	stateByInstanceName := make(map[string]*SnapState, len(snapStates))
+	for _, hint := range hints {
+		if _, ok := snapStates[hint.SnapSetup.InstanceName()]; !ok {
+			return nil, nil, snap.NotInstalledError{Snap: hint.SnapSetup.InstanceName()}
```

Comment:
I am a bit unsure about this behavior, shouldn't we simply instead filter out the snaps that don't exist anymore?

Reply:
Indeed, makes sense.